Save test binaries

leandro-lucarella-sociomantic opened the pull request:
When running tests, since the final target is a `test.stamp` instead of producing the binary (because we care about running them, not just building them), the test binaries themselves are considered intermediate files by Make, and normally removed automatically (even if a test failed).

To avoid this, we need to mark them as `.PRECIOUS:`, for example for integration tests:
```make
.PRECIOUS: $(patsubst $C/integrationtest/%/main.d,$O/test-%,$(wildcard $C/integrationtest/*/main.d))
```

mathias-baumann-sociomantic commented:
![image](https://user-images.githubusercontent.com/2702644/40614915-f7bf0018-6284-11e8-9154-86129dffb7a4.png)

leandro-lucarella-sociomantic committed:
Save important generated test files

In case of further debugging is needed, or a re-run of the tests for
other reasons, is good to make sure these generated binaries are kept.
Normally they are built to fulfill the bin.stamp target that runs the
binary, so the binary itself is considered an intermediate files.

For more information, see
https://www.gnu.org/software/make/manual/html_node/Chained-Rules.html#Chained-Rules

Fixes #124.

leandro-lucarella-sociomantic commented:
Attached fix.

leandro-lucarella-sociomantic commented:
I guess you didn't actually test it, right? I was about to do a simple test.

leandro-lucarella-sociomantic commented:
Usually when I assign a PR to myself is because it's ready to be reviewed (and I appreciate approvals) but I want to do the merging myself (because I'm checking something for example).

leandro-lucarella-sociomantic commented:
My simple test worked, all seems :heavy_check_mark:

mathias-baumann-sociomantic commented:
I didn't test, no. Will only "approve" next time then :)
